Consume messages from stream-company-profile and post to upsert endpoint in search.api.ch.gov.uk

caismith-ch opened the pull request:
- Add config required to consume from kafka2 topics i.e `stream-company-profile`
- Add main file which creates consumer/consumer group
- Implement service which deals with the consuming of messages
- Add REST request to search.api.ch.gov.uk to upsert companies from the `stream-company-profile` topic. This means that if the company exists in the index it will updated, if it doesn't it will be added to the index.
- Add make, start and version files
- Add basic Readme file
- Add unit tests and mocks

Resolves: GCI-411

caismith-ch committed:
Add Makefile, start script and version file for initial commit

caismith-ch committed:
Add main.go and config files

caismith-ch committed:
Add code to start and log the consumer

caismith-ch committed:
Update config with additional environment variables that will be required and create consumer group

caismith-ch committed:
Remove commented out code and update config to include consumer group

caismith-ch committed:
Update configs to use streaming versions i.e streaming zookeeper

caismith-ch committed:
Add basic README to provide context to the service and basic requirements and tips

jsimons1989 commented on Makefile:
```diff
@@ -0,0 +1,67 @@
+
+commit        := $(shell git rev-parse --short HEAD)
+tag           := $(shell git tag -l 'v*-rc*' --points-at HEAD)
+version       := $(shell if [[ -n "$(tag)" ]]; then echo $(tag) | sed 's/^v//'; else echo $(commit); fi)
```
These variables aren't needed. The version will be read from the version file.

jsimons1989 commented on Makefile:
```diff
@@ -0,0 +1,67 @@
+TESTS         ?= ./...
+
+bin           := company-search-consumer
+test_path     := ./test
```
This variable isn't being used

jsimons1989 commented on Makefile:
```diff
@@ -0,0 +1,67 @@
+
+.PHONY: test-unit
+test-unit:
+	go test $(TESTS) -run 'Unit'
```
This needs a flag to record the coverage for Sonar

jsimons1989 commented on README.md:
```diff
@@ -1 +1,23 @@
-# company-search-consumer
\ No newline at end of file
+# company-search-consumer
+A service that consumes messages from the stream-company-profile topic and then sends a RESTful POST request to the search.api.ch.gov.uk service. The data sent through the POST request will be used in search.api.ch.gov.uk to maintain and update the elastic search alpha_index.
```
This summary will be incorrect if the functionality of search.api.ch.gov.uk changes.

jsimons1989 commented on README.md:
```diff
@@ -1 +1,23 @@
+- [Go](https://golang.org/doc/install)
+- [Git](https://git-scm.com/downloads)
+- [Kafka](https://kafka.apache.org/) - this is installed as part of the chs local build (within the chs-kafka VM)
+- [Kafka 2]
```
This link is incomplete

jsimons1989 commented on README.md:
```diff
@@ -1 +1,23 @@
+# company-search-consumer
+A service that consumes messages from the stream-company-profile topic and then sends a RESTful POST request to the search.api.ch.gov.uk service. The data sent through the POST request will be used in search.api.ch.gov.uk to maintain and update the elastic search alpha_index.
+
+Requirements (local setup)
```
The requirements should be the same for a dev and live setup.

jsimons1989 commented on README.md:
```diff
@@ -1 +1,23 @@
+Getting started
+--------------
+To build the service:
+ 1. Clone the repository into your GOPATH under `src/github.com/companieshouse`
```
The repo no longer needs to be in the GOPATH when using Go modules

jsimons1989 commented on main.go:
```diff
@@ -0,0 +1,49 @@
+	log.Namespace = "company-search-consumer"
+	if len(os.Getenv("HUMAN_LOG")) < 1 {
+		klg := &klog.Klog{OutTopic: cfg.LogTopic, BrokerAddr: cfg.StreamingBrokerAddr, SchemaRegistryURL: cfg.SchemaRegistryURL}
+		go klg.Capture()
```
Is the klog block still required? I don't think this is needed.

jsimons1989 commented on main.go:
```diff
@@ -0,0 +1,49 @@
+	if err := consumer.JoinGroup(groupConfig); err != nil {
+		log.ErrorC("Error joining consumer group", err, nil)
+		return
+	}
```
I would advise adding a log message to notify that the consumer has been successfully initialised

jsimons1989 commented on main.go:
```diff
@@ -0,0 +1,49 @@
+		Chroot:      cfg.ZookeeperChroot,
+	}
+
+	log.Debug("creating consumer group")
```
This doesn't output which consumer group.
e.g.
```
log.Debug("Joining consumer group", log.Data{"group": cfg.ConsumerGroupName});
```

jsimons1989 commented on main.go:
```diff
@@ -0,0 +1,49 @@
+		log.ErrorC("Error joining consumer group", err, nil)
+		return
+	}
+}
```
Gracefully shutting down the consumer when the application is terminated is missing
```
log.Info("Closing consumer", log.Data{"topic": topic})
err = svc.Consumer.Close()
if err != nil {
    log.Error(fmt.Errorf("Error closing consumer: %s", err))
}
log.Info("Consumer successfully closed", log.Data{"topic": svc.Topic})
```

caismith-ch replied:
Currently the consumer isn't polling and there is no service. To prevent the PR from getting two big i've began working on the service code here https://github.com/companieshouse/company-search-consumer/compare/feature/initial-consumer...add-service-functionality and I think the above code would be relevant to this branch instead, what do you think?

caismith-ch replied:
Merged the above branch into this PR to make keeping track easier for myself.

caismith-ch committed:
Address PR Comments:

- Remove klog block of code as no longer relevant
- Update README
- Increase logging accuracy

caismith-ch committed:
Initial commit - add exisiting service from email-sender to adjust to fit company-search needs

caismith-ch committed:
Remove email sender code

caismith-ch committed:
Commit code at point in which I can prove that the consumer is polling for messages from stream-company-profile

caismith-ch committed:
Successfully unmarshal the kafka message the data can then be used to post to search api

caismith-ch committed:
Add ability to upsert a company to the elastic search alpha_index

caismith-ch committed:
Merge pull request #2 from companieshouse/add-service-functionality

Add service functionality

caismith-ch committed:
Add test for upsert functionality

caismith-ch committed:
Initial attempt at testing the service code

caismith-ch committed:
Add service mock file

caismith-ch committed:
Add service tests and refactor code to use interfaces to provide ability to mock dependancies

caismith-ch committed:
Remove 'replace' in go mod which was used for local testing

caismith-ch committed:
Merge pull request #3 from companieshouse/service-unit-test

Add unit tests and refactor code to use interfaces to provide ability to mock dependencies

BSaunders95 commented on main.go:
```diff
@@ -0,0 +1,78 @@
+
+	upsert := &upsert.Template{
+		HTTPClient:          http.DefaultClient,
+		UpsertCompanyAPIUrl: "http://api.chs-dev.internal:4089/upsert-company",
```
Should this be configurable?

caismith-ch replied:
Definitely!

jsimons1989 commented on README.md:
```diff
@@ -1 +1,23 @@
-# company-search-consumer
\ No newline at end of file
+# company-search-consumer
+A service that consumes messages from the stream-company-profile topic and then sends a RESTful POST request to the search.api.ch.gov.uk service. The data sent through the POST request will be used in search.api.ch.gov.uk to maintain and update elastic search indexes relating to company search.
```
The last sentence of this summary is stating the behaviour of a different service. If the other service changes, this documentation will be incorrect. Can this summary be a high level explanation of the purpose of this consumer only please?

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+)
+
+// Value required to unit test the consumer
+var testHttpResponseValue int
```
There shouldn't be a test variable in non-test code.

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+type Event struct {
+	FieldsChanged []string `avro:"fields_changed,omitempty"`
+	PublishedAt   string   `avro:"published_at"`
+	Type          string   `avro:"type"`
```
These are already defined in [chs-go-avro-schemas](https://github.com/companieshouse/chs-go-avro-schemas)

emead replied:
This has not yet been resolved?

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+					continue
+				}
+
+				if err != nil {
```
There's two error checks in a row. The second one will never be called?

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+					continue
+				}
+
+				log.Event("trace", "", log.Data{"data": mm.Data})
```
Why is an empty string being passed into the log? Assuming this is a trace message, can't the trace function be used?
```
log.Trace(..)
```

A data object shouldn't be output into the log. Please output the required fields only.

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+				}
+
+				log.Event("trace", "", log.Data{"data": mm.Data})
+				statusCode, err := svc.Upsert.SendViaAPI(mm.Data)
```
The consumer shouldn't know about the use of HTTP. This makes the code more difficult to maintain.

jsimons1989 commented on service/service.go:
```diff
@@ -0,0 +1,102 @@
+
+				svc.Consumer.MarkOffset(event, "")
+				if err := svc.Consumer.CommitOffsets(); err != nil {
+					log.ErrorC("Error committing message offset", err, nil)
```
Is the offset which failed to be committed going to be in the err? This information might make diagnosing and error resolution quicker to resolve.

jsimons1989 commented on service/service_mocks.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (m MockConsumer) Errors() <-chan error {
+	return nil
```
The handling of errors should be tested?

jsimons1989 commented on service/service_mocks.go:
```diff
@@ -0,0 +1,85 @@
+package service
```
Only `mock_*.go` and `*_mock.go` are being excluded from Sonar coverage. This will drop the test coverage stats.

jsimons1989 commented on service/service_mocks.go:
```diff
@@ -0,0 +1,85 @@
+
+type MockUpsertFail struct{}
+
+type MockGroup struct{}
```
This is a lot of mocking in a single unit. Maybe this service is doing too much? Could the consumer be separated from the service so the consumer only worries about consuming from Kafka and the service handles the processing?

jsimons1989 commented on service/service_mocks.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (m MockUpsertFail) SendViaAPI(data string) (int, error) {
+	return 400, errors.New("failed to send to api")
```
A single mock could be used for the different test cases to simplify the testing.
```
type MockUpsert struct {
   Code int
   Error error
}

func (m MockUpsert) SendViaAPI(data string) (int, error) {
    return m.Code, m.Error
}
```

jsimons1989 commented on service/service_test.go:
```diff
@@ -0,0 +1,78 @@
+
+				svc.Upsert = MockUpsert{}
+
+				Convey("Then the consumer will close succesfully", func() {
```
The `Then` clause should be asserting the correct behaviour.
These tests are a bit misleading as the service doesn't terminate the consumer after a message is processed.

jsimons1989 commented on service/service_test.go:
```diff
@@ -0,0 +1,78 @@
+	}
+}
+
+func TestUnitStart(t *testing.T) {
```
This missing test coverage for all of the failure scenarios, e.g. failing to unmarshal the Avro.

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// Template config contains recipient address, http client and send email endpoint
+type Template struct {
```
This struct contains the implementations of the interface. Why isn't it called APIUpsert or HTTPUpsert to match the interface name?

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+var (
+	//ErrInvalidResponse is returned when the Versioning API returns a non 200 status
```
Should be "search API" not "Versioning API" in the comment

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+	ErrInvalidResponse = errors.New("invalid status returned by search.api.ch.gov.uk")
+)
+
+type Upsert interface {
```
Please add a comment to this interface

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+
+// SendViaAPI makes a call to the search.api.ch.gov.uk and passes it
+// the required data for upserting to alpha_search index
+func (upsert *Template) SendViaAPI(data string) (int, error) {
```
Shouldn't this use the SDK?

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+// the required data for upserting to alpha_search index
+func (upsert *Template) SendViaAPI(data string) (int, error) {
+
+	req, err := http.NewRequest("POST", upsert.UpsertCompanyAPIUrl, strings.NewReader(data))
```
Why is `POST` being used for an upsert? Shouldn't this be a `PUT`?

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+
+	req, err := http.NewRequest("POST", upsert.UpsertCompanyAPIUrl, strings.NewReader(data))
+	if err != nil {
+		return 400, err
```
This shouldn't assume the error code is a 400. There isn't a status code yet?

jsimons1989 commented on upsert/upsert.go:
```diff
@@ -0,0 +1,45 @@
+
+	resp, err := upsert.HTTPClient.Do(req)
+	if err != nil {
+		return 400, err
```
This shouldn't assume the error code is a 400

jsimons1989 commented on upsert/upsert_test.go:
```diff
@@ -0,0 +1,61 @@
+
+		upsert := &Template{
+			HTTPClient:          httpClient,
+			UpsertCompanyAPIUrl: "http://api.chs-dev.internal:4089/upsert-company",
```
This shouldn't expose a real URL

jsimons1989 commented on upsert/upsert_test.go:
```diff
@@ -0,0 +1,61 @@
+		}
+		statusCode, err := upsert.SendViaAPI("{ 'data' : '1' }")
+		So(err, ShouldNotBeNil)
+		So(statusCode, ShouldEqual, 400)
```
This status code doesn't match the status code in the createMockClient() function call?

jsimons1989 reviewed (dismissed):
Please resolve the comments

jsimons1989 commented:
Why have more changes been added to this PR when it's already been through a review? A new PR should have been opened for new code.

caismith-ch committed:
- Refactor README to not be dependant on another service

- Rename variables to better describe what they do or are for
- Remove unnessary if statement and log message
- Change file to be 'mock' instead of 'mocks' for sonar exclusions
- Add config for upsert url and implement it

Commit:
Updated unit tests and refactored upsertAPI code

Commit:
Changed POST request to PUT request

emead commented on config/config.go:
```diff
@@ -0,0 +1,47 @@
+	LogTopic                  string      `env:"LOG_TOPIC"                           flag:"log-topic"                           flagDesc:"Log topic"`
+	UpsertURL                 string      `env:"UPSERT_URL"                          flag:"upsert-url"                          flagDesc:"Url for upserting a company"`
+	RetryThrottleRate         int         `env:"RETRY_THROTTLE_RATE_SECONDS"         flag:"retry-throttle-rate-seconds"         flagDesc:"Retry throttle rate seconds"`
+	MaxRetryAttempts          int         `env:"MAXIMUM_RETRY_ATTEMPTS"              flag:"max-retry-attemps"                   flagDesc:"Maximum retry attempts"`
```
Spelling: `attemps`

emead commented on config/config.go:
```diff
@@ -0,0 +1,47 @@
+import (
+	"github.com/companieshouse/chs.go/log"
+	"github.com/companieshouse/gofigure"
+	"os"
```
`goimports` should be run against every file

emead commented on config/config.go:
```diff
@@ -0,0 +1,47 @@
+package config
```
Package should have a comment

emead commented on main.go:
```diff
@@ -0,0 +1,92 @@
+		os.Exit(1)
+	}
+
+	p, err := producer.New(&producer.Config{Acks: &producer.WaitForAll, BrokerAddrs: cfg.StreamingBrokerAddr})
```
`err` should be handled

emead commented on main.go:
```diff
@@ -0,0 +1,92 @@
+		cfg.MaxRetryAttempts,
+	}
+
+	ppSchema, err := schema.Get(cfg.SchemaRegistryURL, resourceChangedDataSchema)
```
`err` should be handled

emead commented on main.go:
```diff
@@ -0,0 +1,92 @@
+	retry := &resilience.ServiceRetry{
+		time.Duration(cfg.RetryThrottleRate),
+		cfg.MaxRetryAttempts,
+	}
```
Keys should be named
```suggestion
      retry := &resilience.ServiceRetry{
		ThrottleRate: time.Duration(cfg.RetryThrottleRate),
		MaxRetries:   cfg.MaxRetryAttempts,
	}
```

emead commented on main.go:
```diff
@@ -0,0 +1,92 @@
+
+	log.Debug("Consumer has been successfully initialised")
+
+	avro := &avro.Schema{
```
Variable names must not be the same as imported package names (same applies for consumer, avro, upsert)

emead commented on service/service.go:
```diff
@@ -0,0 +1,92 @@
+package service
```
Should have a package comment

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+package upsert
```
Should have a package comment

emead commented on service/service.go:
```diff
@@ -0,0 +1,92 @@
+	Schema        string
+}
+
+//Start is called to run the service
```
Please put a space at the start of the comment
```suggestion
// Start is called to run the service
```

emead commented on service/service.go:
```diff
@@ -0,0 +1,92 @@
+			log.Info("Closing consumer")
+			err := svc.Consumer.Close()
+			if err != nil {
+				log.Error(fmt.Errorf("Error closing consumer: %s", err))
```
Error strings should not begin with a capital letter
```suggestion
				log.Error(fmt.Errorf("error closing consumer: %s", err))
```

emead commented on service/service.go:
```diff
@@ -0,0 +1,92 @@
+
+				if err := svc.Upsert.SendViaAPI(mm.Data); err != nil {
+					log.ErrorC("Error calling upsert on the search api", err, nil)
+					svc.HandleError(err, message.Offset, &mm)
```
`HandleError` returns an error which must be handled (in all places where `HandleError` is called.

emead commented on service/service_test.go:
```diff
@@ -0,0 +1,260 @@
+	}
+}
+
+func createMockServiceErrorCommitting(mockUpsert *upsert.MockUpsert) *Service {
```
This function appears to be unused and should be deleted

emead commented on service/service_test.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+const resourceKind = "resourceKind"
+const resourceUri = "resourceUri"
```
Rename to `resourceURI`

emead replied:
Not yet done

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+var (
+	//ErrInvalidResponse is returned when the Search API returns a non 200 status
```
Please put a space at the start of the comment
```suggestion
	// ErrInvalidResponse is returned when the Search API returns a non 200 status
```

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+	ErrInvalidResponse = errors.New("invalid status returned by search.api.ch.gov.uk")
+)
+
+// Interface for calling api to upsert data to an elastic search index
```
Comment must begin with interface name
```suggestion
// Upsert is an interface for calling API to upsert data to an elastic search index
```

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+
+// Interface for calling api to upsert data to an elastic search index
+type Upsert interface {
+	// SendViaApi provides ability to post to search api
```
Casing incorrect
```suggestion
	// SendViaAPI provides ability to post to search api
```

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode != 200 {
```
Safer to use status code
```suggestion
	if resp.StatusCode != http.StatusOK {
```

emead commented on upsert/upsert.go:
```diff
@@ -0,0 +1,46 @@
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
```
Error return value not checked

emead commented:
Please add a .gitignore file

Commit:
Ran goimports

Commit:
Added comments and fixed imports, added more error handling

Commit:
Fixed typo in config

jphillips2121 committed:
Added chs-go-avro-schema dependency

jphillips2121 committed:
Updated chs.go version to 1.1.5

jphillips2121 committed:
Renamed resourceURI variable
